File: app/src/main/java/com/sunmoon/helper/modules/helper/HelperViewModel.java
package com.sunmoon.helper.modules.helper;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.os.Build;
import android.os.Bundle;
import android.speech.RecognitionListener;

import com.sunmoon.helper.api.ApiManage;
import com.sunmoon.helper.base.BaseViewModel;
import com.sunmoon.helper.common.Flag;
import com.sunmoon.helper.modules.remind.EditRemindActivity;
import com.sunmoon.helper.modules.search.SearchActivity;
import com.sunmoon.helper.model.Message;
import com.sunmoon.helper.model.Phone;
import com.sunmoon.helper.model.TuLing;
import com.sunmoon.helper.model.UserCommand;
import com.sunmoon.helper.utils.Apk;

import sunmoon.voice.recognition.BaiduUtil;

import com.sunmoon.helper.utils.PhoneUtil;
import com.sunmoon.helper.utils.StringUtil;
import com.sunmoon.helper.utils.UserPurpose;

import java.util.List;

import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import sunmoon.voice.recognition.SpeechOcr;
import sunmoon.voice.recognition.VoiceWakeUp;
import sunmoon.voice.recognition.WakeUpListener;
import sunmoon.voice.tts.Tts;


/**
 * Created by SunMoon on 2016/11/30.
 */

@TargetApi(Build.VERSION_CODES.FROYO)
public class HelperViewModel extends BaseViewModel implements RecognitionListener  {
    private UserPurpose userPurpose;
    private VoiceAdapter adapter;
    private List<PackageInfo> packageInfos;
    private List<Phone> phoneInfos;
    private Context context;
    private Tts tts = new Tts();
    private VoiceWakeUp wakeUp;
    private final int REC_CODE = 1;// 语音识别
    private ChatView v;
    public HelperViewModel(final Context context) {
        this.userPurpose = new UserPurpose();
        this.context = context;
        adapter = new VoiceAdapter(context);
        tts.init(context);
        wakeUp = new VoiceWakeUp(context);
        wakeUp.registerListener(new WakeUpListener() {
            @Override
            public void success() {
                startRecEvent();
            }

            @Override
            public void exit() {

            }
        });
        SpeechOcr.setRecognitionListener(this);
    }
    public void setView(ChatView v){
        this.v = v;
    }
    public VoiceAdapter getAdapter() {
        return adapter;
    }

    public void startWakeUp() {
        wakeUp.start();
    }

    public void startRecEvent() {
        tts.stop();
        wakeUp.stop();
        v.startRecDialog(SpeechOcr.startDialogRec(), REC_CODE);
    }

    public void stopAll() {
        SpeechOcr.stopAll();
    }

    @Override
    public void onReadyForSpeech(Bundle bundle) {

    }

    @Override
    public void onBeginningOfSpeech() {

    }

    @Override
    public void onRmsChanged(float v) {
        this.v.onRmsChanged(v);
    }

    @Override
    public void onBufferReceived(byte[] bytes) {

    }

    @Override
    public void onEndOfSpeech() {

    }

    @Override
    public void onError(int i) {

    }

    @Override
    public void onResults(Bundle bundle) {
        String raw = BaiduUtil.getRecResult(bundle);
        String result = StringUtil.ridPunctuation(raw);
        sendRightMsg(result);
        handleResult(userPurpose.getUserCommand(result));

    }

    public void sendLeftMsg(String msg, boolean isSpeak) {
        Message msgs = new Message(msg, 0);
        adapter.addMessage(msgs);
        if (isSpeak) {
            tts.speak(msg);
        }
        v.smoothBottom();
    }

    public void sendRightMsg(String msg) {
        v.smoothBottom();
        adapter.addMessage(new Message(msg, 1));
    }

    public void handleResult(UserCommand userCommand) {
        String target = userCommand.getSentence().getTarget();
        switch (userCommand.getType()) {
            case UserCommand.COMMAND_CALL_PHONE:
                if (phoneInfos == null || phoneInfos.size() == 0) {
                    phoneInfos = PhoneUtil.getPhoneInfos(context);
                }
                List<Phone> result = PhoneUtil.getPhonesByName(phoneInfos, target);
                if (result.size() > 0) {
                    sendLeftMsg("正在为您呼叫: " + result.get(0).getName(), false);
                    PhoneUtil.fastCallPerson(context, result.get(0).getNumber());
                } else {
                    sendLeftMsg("未找到联系人:" + target, true);
                }
                break;
            case UserCommand.COMMAND_OPEN_APP:
                if (packageInfos == null || packageInfos.size() == 0) {
                    packageInfos = Apk.getPackagesInfos(context);
                }
                String packName = Apk.getAppPackageName(context, packageInfos, target);
                if (!packName.equals(Flag.FAIL)) {
                    sendLeftMsg("正在打开：" + target, false);
                    Apk.openApp(context, packName);
                } else {
                    sendLeftMsg("未找到应用：" + target, true);
                }
                break;
            case UserCommand.COMMAND_DELETE_APP:
                if (packageInfos == null || packageInfos.size() == 0) {
                    packageInfos = Apk.getPackagesInfos(context);
                }
                packName = Apk.getAppPackageName(context, packageInfos, target);
                if (!packName.equals(Flag.FAIL)) {
                    sendLeftMsg("正在卸载：" + target, false);
                    Apk.unInstallApp(context, packName);

                } else {
                    sendLeftMsg("未找到应用：" + target, true);
                }
                break;
            case UserCommand.COMMAND_SEARCH:
                Intent intent = new Intent(context,SearchActivity.class);
                intent.putExtra("keyword", target);
                context.startActivity(intent);
                break;
            case UserCommand.COMMAND_REMIND:
                 intent = new Intent(context, EditRemindActivity.class);
                break;
            case UserCommand.COMMAND_CHAT:
                Subscription rx = ApiManage.getInstence().getTuLingService().getResult(com.sunmoon.helper.common.TuLing.API_KEY, target)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(new Action1<TuLing>() {
                            @Override
                            public void call(TuLing tuLing) {
                                sendLeftMsg(tuLing.getText(), true);
                            }
                        }, new Action1<Throwable>() {
                            @Override
                            public void call(Throwable throwable) {
                                sendLeftMsg(throwable.getMessage(), false);
                            }
                        });
                break;
        }
    }

    @Override
    public void onPartialResults(Bundle bundle) {

    }

    @Override
    public void onEvent(int i, Bundle bundle) {

    }


    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REC_CODE && data != null) {
            onResults(data.getExtras());
        }
    }

}

File: app/src/main/java/com/sunmoon/helper/modules/helper/HelperViewModel.kt
package com.sunmoon.helper.modules.helper

import android.annotation.TargetApi
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageInfo
import android.databinding.ObservableField
import android.os.Build
import android.widget.ImageView
import cn.jiguang.imui.commons.ImageLoader
import cn.jiguang.imui.commons.models.IMessage
import cn.jiguang.imui.messages.MsgListAdapter
import com.bumptech.glide.Glide
import com.sunmoon.helper.api.ApiManage
import com.sunmoon.helper.base.BaseViewModel
import com.sunmoon.helper.common.Flag
import com.sunmoon.helper.model.DefaultUser
import com.sunmoon.helper.model.MyMessage
import com.sunmoon.helper.model.Phone
import com.sunmoon.helper.model.UserCommand
import com.sunmoon.helper.modules.remind.EditRemindActivity
import com.sunmoon.helper.modules.search.SearchActivity
import com.sunmoon.helper.utils.Apk
import com.sunmoon.helper.utils.PhoneUtil
import com.sunmoon.helper.utils.UserPurpose
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import sunmoon.voice.control.SpeechRec
import sunmoon.voice.tts.Speaker

/**
 * Created by SunMoon on 2016/11/30.
 */

@TargetApi(Build.VERSION_CODES.FROYO)
class HelperViewModel() : BaseViewModel() {
    private var curUserMsg: MyMessage? = null
    private var mPhoneList: List<Phone>? = null
    private var mPackageList: List<PackageInfo>? = null
    var adapter: MsgListAdapter<MyMessage>? = null
    val tip = ObservableField<String>("点击开始识别")
    val isSpeechModel = ObservableField<Boolean>(true)
    override fun init(context: Activity) {
        super.init(context)
        adapter = MsgListAdapter("0", object : ImageLoader {
            override fun loadAvatarImage(avatarImageView: ImageView?, string: String?) {
                Glide.with(context).load(string).into(avatarImageView)
            }
            override fun loadImage(imageView: ImageView?, string: String?) {
                Glide.with(context).load(string).into(imageView)
            }
        })
        mPhoneList = PhoneUtil.getPhones(context)
        mPackageList = Apk.getPackagesInfos(context)
    }
    fun start() {
        SpeechRec.getInstance().start()
    }
    fun sendMsg(msg: String) {
        val msg = MyMessage(msg, IMessage.MessageType.SEND_TEXT)
        msg.status = IMessage.MessageStatus.CREATED
        msg.setUserInfo(DefaultUser("1", "yonghu", "file:///android_asset/my_header.jpg"))
        curUserMsg = msg
        adapter?.addToStart(msg, true)
    }
    private fun replyMsg(msg: String) {
        if (curUserMsg != null) {
            curUserMsg?.status = IMessage.MessageStatus.SEND_SUCCEED
            adapter?.updateMessage(curUserMsg)
        }
        adapter?.addToStart(MyMessage(msg, IMessage.MessageType.RECEIVE_TEXT), true)
    }
    fun execCommand(content: String) {
        sendMsg(content)
        val userCommand = UserPurpose.create(content)
        val target = userCommand.sentence.target
        when (userCommand.type) {
            UserCommand.COMMAND_CALL_PHONE  // 打电话
            -> {
                val result = PhoneUtil.getPhonesByName(mPhoneList, target)
                if (result.size > 0) {
                    replyMsg("正在为您呼叫: " + result[0].name)
                    if (!PhoneUtil.fastCallPerson(context, result[0].number)){
                        PhoneUtil.callPerson(context,result[0].number)
                    }
                } else {
                    replyMsg("未找到联系人:" + target)
                }
            }
            UserCommand.COMMAND_OPEN_APP// 打开APP
            -> {
                var packName = Apk.getAppPackageName(context, mPackageList, target)
                if (packName != Flag.FAIL) {
                    replyMsg("正在打开：" + target)
                    Apk.openApp(context, packName)
                } else {
                    replyMsg("未找到应用：" + target)
                }
            }
            UserCommand.COMMAND_DELETE_APP // 卸载APP
            -> {
                val packName = Apk.getAppPackageName(context, mPackageList, target)
                if (packName != Flag.FAIL) {
                    replyMsg("正在卸载：" + target)
                    Apk.unInstallApp(context, packName)

                } else {
                    replyMsg("未找到应用：" + target)
                }
            }
            UserCommand.COMMAND_SEARCH // 网络搜索
            -> {
                replyMsg("成功为您搜索："+ target)
                var intent = Intent(context, SearchActivity::class.java)
                intent.putExtra("keyword", target)
                context?.startActivity(intent)
            }
            UserCommand.COMMAND_REMIND // 添加提醒
            -> {
                val intent = Intent(context, EditRemindActivity::class.java)
                intent.putExtra("isNew", true)
                intent.putExtra("remindContent", userCommand.sentence.target)
                intent.putExtra("similarTime", userCommand.sentence.qualifier)
                context?.startActivity(intent)
            }
            UserCommand.COMMAND_CHAT -> {
                val rx = ApiManage.getInstence().tuLing.getResult(com.sunmoon.helper.common.TuLing.API_KEY, target)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ tuLing ->
                            replyMsg(tuLing.text)
                            if (HelperSetting.canSpeak) {
                                Speaker.getInstance().start(tuLing.text)
                            }
                        }) { throwable ->
                            replyMsg(throwable.message!!)

                        }

            }
        }
    }
}
